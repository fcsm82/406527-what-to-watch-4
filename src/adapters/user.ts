import {BASE_URL} from "../consts";

export default class User {
  id: number;
  email: string;
  name: string;
  avatarSrc: string;

  constructor(user) {
    this.id = user.id;
    this.email = user.email;
    this.name = user.name;
    this.avatarSrc = `${BASE_URL}${user.avatar_url}`;
  }

  static parse(user) {
    return Object.keys(user).length ? new User(user) : {};
  }
}
