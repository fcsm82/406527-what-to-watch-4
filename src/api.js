import axios from 'axios';
import history from './history';

import {BASE_URL, AppRoute} from "./consts";

const TIMEOUT = 5000;
const URL = `${BASE_URL}/wtw`;

const Error = {
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  NOT_FOUND: 404,
};

export const createAPI = (onUnauthorized) => {
  const api = axios.create({
    baseURL: URL,
    timeout: TIMEOUT,
    withCredentials: true,
  });

  const onSuccess = (response) => {
    return response;
  };

  const onFail = (err) => {
    const {response, request} = err;

    switch (response.status) {
      case Error.UNAUTHORIZED:
        if (request.responseURL !== URL + `/login`) {
          history.push(AppRoute.LOGIN);
        }
        break;

      default:
        throw err;
    }

    onUnauthorized();
    throw err;
  };

  api.interceptors.response.use(onSuccess, onFail);

  return api;
};
