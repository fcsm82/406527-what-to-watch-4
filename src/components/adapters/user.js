import {BASE_URL} from "../../consts.js";

export default class User {
  constructor(user) {
    this.id = user.id;
    this.email = user.email;
    this.name = user.name;
    this.avatarSrc = `${BASE_URL}/${user.avatar_url}`;
  }

  static parse(user) {
    return Object.keys(user).length ? new User(user) : {};
  }
}
