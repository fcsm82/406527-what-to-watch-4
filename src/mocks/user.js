export const user = {
  id: 1213,
  email: `john.doe@johndoe.com`,
  name: `John Doe`,
  avatarSrc: `avatarSrc`,
};
