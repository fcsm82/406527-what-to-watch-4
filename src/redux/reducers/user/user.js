import {Action, requireAuthorization} from "../../actions.js";

export const AuthorizationStatus = {
  AUTH: `Auth`,
  NO_AUTH: `No auth`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case Action.REQUIRED_AUTHORIZATION:
      return Object.assign({}, state, {
        authorizationStatus: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then(() => {
        dispatch(requireAuthorization(AuthorizationStatus.AUTH));
      });
  },
};


export {reducer, Operation};
