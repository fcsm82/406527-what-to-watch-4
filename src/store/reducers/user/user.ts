import {Action, Creator} from "./actions";
import history from "../../../history";

import User from "../../../adapters/user";

import {AppRoute} from "../../../consts";

export enum AuthorizationStatus {
  AUTH = `Auth`,
  NO_AUTH = `No auth`,
}

const initialState = {
  user: {},
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authorizationError: false,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case Action.SET_AUTHORIZATION_STATUS:
      return Object.assign({}, state, {
        authorizationStatus: action.payload,
      });

    case Action.SET_AUTHORIZATION_ERROR:
      return Object.assign({}, state, {
        authorizationError: action.payload,
      });

    case Action.SET_USER_DATA:
      return Object.assign({}, state, {
        user: action.payload,
      });

    default:
      return state;
  }
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(AppRoute.LOGIN)
      .then((response) => {
        dispatch(Creator.setAuthorizationStatus(AuthorizationStatus.AUTH));
        dispatch(Creator.setUserData(User.parse(response.data)));
      })
      .catch((err) => {
        dispatch(Creator.setAuthorizationStatus(AuthorizationStatus.NO_AUTH));
        dispatch(Creator.setAuthorizationError(err));
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(AppRoute.LOGIN, {
      email: authData.login,
      password: authData.password,
    })
    .then((response) => {
      dispatch(Creator.setAuthorizationStatus(AuthorizationStatus.AUTH));
      dispatch(Creator.setUserData(User.parse(response.data)));
      history.push(AppRoute.ROOT);
    })
    .catch((err) => {
      dispatch(Creator.setAuthorizationStatus(AuthorizationStatus.NO_AUTH));
      dispatch(Creator.setAuthorizationError(err));
    });
  }
};

export {reducer, Operation};
