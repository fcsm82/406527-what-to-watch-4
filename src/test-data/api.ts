import axios from 'axios';

export const api = axios.create({
  baseURL: `https://4.react.pages.academy/wtw`,
  timeout: 5000,
  withCredentials: true,
});
