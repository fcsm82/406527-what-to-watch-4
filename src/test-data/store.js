import NameSpace from "../store/reducers/name-space.js";
import {promoFilm, films} from "../test-data/films.js";
import {user} from "../test-data/user.js";

export const testStore = {
  [NameSpace.DATA]: {
    isAppLoading: true,
    isAppError: false,
    films,
    promoFilm,
    favoriteFilms: [],
    comments: []
  },
  [NameSpace.FILMS]: {
    shownFilms: 8,
    activeFilter: `All genres`,
  },
  [NameSpace.USER]: {
    error: null,
    authorizationStatus: `No auth`,
    authorizationError: false,
    user
  },
};

