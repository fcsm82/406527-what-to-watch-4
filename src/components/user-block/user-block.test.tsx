import * as React from "react";
import * as renderer from "react-test-renderer";
import {Router} from "react-router-dom";

import {Provider} from "react-redux";

import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";

import history from "../../history";

import {testStore} from "../../test-data/store";
import UserBlock from "./user-block";
import {api} from "../../test-data/api";
import {user} from "../../test-data/user";

const middleware = [thunk.withExtraArgument(api)];
const mockStore = configureStore(middleware);
const store = mockStore(testStore);

it(`Should UserBlock render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <Provider store={store}>
            <UserBlock
              authorizationStatus = {`No auth`}
              user = {user}
            />
          </Provider>
        </Router>, {
          createNodeMock: () => {
            return {};
          }
        }).toJSON();

  expect(tree).toMatchSnapshot();
});
