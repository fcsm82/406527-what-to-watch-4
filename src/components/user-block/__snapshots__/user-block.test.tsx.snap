// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should UserBlock render correctly 1`] = `
<div
  className="user-block"
>
  <a
    href="/login"
    onClick={[Function]}
    style={
      Object {
        "color": "inherit",
        "cursor": "pointer",
        "textDecoration": "none",
      }
    }
  >
    Sign In
  </a>
</div>
`;
